README.md
# tiny-etag

ETag middleware for express.js API compatible routers.

## Using the factory

```javascript
import {etag} from "tiny-etag";
const etags = etag({cacheSize: 500});
const router = SomeRouter(); /* express.js compatible router */

router.use(etags.middleware);

router.get("/", (req, res) => {
    const body = "Hello World!";

    res.writeHead(200, {"content-type": "text/plain", "etag": etags.create(body)});
    res.end(body);
});
```

## Testing

Tiny ETag has 100% code coverage with its tests.

```console
---------------|---------|----------|---------|---------|---------------------
File           | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
---------------|---------|----------|---------|---------|---------------------
All files      |     100 |    78.57 |     100 |     100 |                    
 tiny-etag.cjs |     100 |    78.57 |     100 |     100 | 45-49,59,91-109,115
---------------|---------|----------|---------|---------|---------------------
```

## API

### etag({cacheSize: 1000, cacheTTL: 0, seed: random, mimetype: "text/plain"})
Returns an tiny-etag instance. Cache TTL concerns do not spread with a notification.

### create(arg = ""[, mimetype = "text/plain"])
Creates a strong ETag value from `arg`; a composite `String` is recommended. It's ideal to 

### hash(arg = ""[, mimetype="text/plain"])
Creates a hash of `arg`, uses `create()`

### keep(arg)
Returns a boolean if `arg` should be kept on the cached `Object`

### middleware(req, res, next)
Middleware to be used by an http framework

### parse(arg)
Parses `arg` as a `URL` if it's a `String`, or constructs one if it is a `socket`

### register(url, state)
Adds `url` to the `cache`

### valid(headers)
Returns a `Boolean` indicating if caching is valid based on `cache-control`

## License
Copyright (c) 2023 Jason Mulligan
Licensed under the BSD-3 license


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [4.0.5](https://github.com/avoidwork/tiny-etag/compare/4.0.5...4.0.5)

#### [4.0.5](https://github.com/avoidwork/tiny-etag/compare/4.0.4...4.0.5)

> 10 October 2023

- Fixing `middleware()` such that it doesn't use extra parameters for `res.send()` [`d6069f8`](https://github.com/avoidwork/tiny-etag/commit/d6069f82f0c370252ce87a2fd8a5cbd376a80079)
- Revert "Decoupling the middleware from a custom decorated method" [`fda1fc2`](https://github.com/avoidwork/tiny-etag/commit/fda1fc26f609b43714fc0cb35268921bdfda25b9)
- Generated CHANGELOG.md [`4a0c531`](https://github.com/avoidwork/tiny-etag/commit/4a0c531b8145193d0be1a05f8f905b1443727a8d)

#### [4.0.4](https://github.com/avoidwork/tiny-etag/compare/4.0.3...4.0.4)

> 10 October 2023

- Caching `content-type` header for 304 responses [`d870760`](https://github.com/avoidwork/tiny-etag/commit/d8707606e50fc997fd95653bfa1527e8ab493f87)
- Generating CHANGELOG.md [`f30b332`](https://github.com/avoidwork/tiny-etag/commit/f30b33252b27ff71fe137da95539bbc7d68262ba)

#### [4.0.3](https://github.com/avoidwork/tiny-etag/compare/4.0.2...4.0.3)

> 10 October 2023

- Decoupling the middleware from a custom decorated method [`688f45e`](https://github.com/avoidwork/tiny-etag/commit/688f45e92d11a2763030f45b0cf0c41f945ec743)
- Generated CHANGELOG.md [`801ddc4`](https://github.com/avoidwork/tiny-etag/commit/801ddc4a601890022fda2b8dbb56e10efb769848)

#### [4.0.2](https://github.com/avoidwork/tiny-etag/compare/4.0.1...4.0.2)

> 10 October 2023

- Bump woodland from 18.0.6 to 18.0.11 [`#15`](https://github.com/avoidwork/tiny-etag/pull/15)
- Bump tiny-lru from 11.2.2 to 11.2.3 [`#14`](https://github.com/avoidwork/tiny-etag/pull/14)
- Bump eslint from 8.50.0 to 8.51.0 [`#13`](https://github.com/avoidwork/tiny-etag/pull/13)
- Bump rollup from 4.0.0 to 4.0.2 [`#12`](https://github.com/avoidwork/tiny-etag/pull/12)
- Bump tiny-httptest from 4.0.5 to 4.0.6 [`#11`](https://github.com/avoidwork/tiny-etag/pull/11)
- Bump woodland from 17.0.2 to 18.0.6 [`#10`](https://github.com/avoidwork/tiny-etag/pull/10)
- Setting extension of `main` to `.cjs` for deprecated tooling [`da21e4e`](https://github.com/avoidwork/tiny-etag/commit/da21e4ef168de72b4b9a3e8ca3a346bd043db12a)
- Generating CHANGELOG.md [`8a8684d`](https://github.com/avoidwork/tiny-etag/commit/8a8684d53e167e87c2c5238c06499e9574995ae1)

#### [4.0.1](https://github.com/avoidwork/tiny-etag/compare/4.0.0...4.0.1)

> 6 October 2023

- Bump rollup from 3.29.4 to 4.0.0 [`#9`](https://github.com/avoidwork/tiny-etag/pull/9)
- Fixing a regression of `hash()` such that the `mimetype` parameter wasn't used in latest iteration, updating README.md, reformatting imports [`ab8f7cf`](https://github.com/avoidwork/tiny-etag/commit/ab8f7cf984e98b1dbbd93a11a5ec1f7ea4fef3f7)

### [4.0.0](https://github.com/avoidwork/tiny-etag/compare/3.0.3...4.0.0)

> 5 October 2023

- Updating dependencies, adding `.github` directory [`e9a003e`](https://github.com/avoidwork/tiny-etag/commit/e9a003e91f5944ca477a53ba93836ea58a145fc8)
- Updating dependencies, adding `.github` directory, WIP [`82e27d2`](https://github.com/avoidwork/tiny-etag/commit/82e27d204822c49204530cfe1d68adc19f3f86a5)
- Updating tests for 100% coverage, updating README.md, adding constants, removing `utils` and moving functions into `ETag` class, setting `engines.node: "&gt;= 17.0.0"` in package.json [`7c9bb7e`](https://github.com/avoidwork/tiny-etag/commit/7c9bb7ea5a1327ba08dd44f73d25cee49ebbc649)

#### [3.0.3](https://github.com/avoidwork/tiny-etag/compare/3.0.2...3.0.3)

> 24 February 2023

- Fixing regression in README.md example from 3.0.0 release, updating LICENSE, version bump to release [`1e51d09`](https://github.com/avoidwork/tiny-etag/commit/1e51d0995132e602a5088d062455cb5c9565905f)
- Generating new CHANGELOG.md [`41342be`](https://github.com/avoidwork/tiny-etag/commit/41342be234d04d347a74274b8fba758a9d75cb81)

#### [3.0.2](https://github.com/avoidwork/tiny-etag/compare/3.0.1...3.0.2)

> 27 October 2022

- Updating dependencies [`2417c88`](https://github.com/avoidwork/tiny-etag/commit/2417c88e6f332b39366fdd64904384b057738a26)
- Updating CHANGELOG.md [`75411a0`](https://github.com/avoidwork/tiny-etag/commit/75411a0201a63234f353b576ecb8f9b491dc0052)

#### [3.0.1](https://github.com/avoidwork/tiny-etag/compare/3.0.0...3.0.1)

> 24 September 2022

- Fixing paths in `package.json` [`7862a0b`](https://github.com/avoidwork/tiny-etag/commit/7862a0bf44f85771d4c209c64cc01a45eafd3613)
- Updating CHANGELOG.md [`e473a99`](https://github.com/avoidwork/tiny-etag/commit/e473a99008c36bc460ec7bb3732b6352c746bf2f)
- Updating the LICENSE file [`0b32a92`](https://github.com/avoidwork/tiny-etag/commit/0b32a925e7e6279d0b3ac3c0cc191f6f1d6ef7f1)

### [3.0.0](https://github.com/avoidwork/tiny-etag/compare/2.1.1...3.0.0)

> 21 September 2022

- Refactoring to an ES module [`#8`](https://github.com/avoidwork/tiny-etag/pull/8)
- Initial changes to an es module [`d15afaf`](https://github.com/avoidwork/tiny-etag/commit/d15afaf73cb7dc34a86ccdff33a461533a46f1bb)
- Implementing esmodule syntax [`0417eb6`](https://github.com/avoidwork/tiny-etag/commit/0417eb6370cc8907e335b2c027bd69b872936880)
- Updating tests, fixing assignment of `mmh3()` [`8d8c6d2`](https://github.com/avoidwork/tiny-etag/commit/8d8c6d2ee9a62102ba2a7a98060c95de9eb5aa7b)

#### [2.1.1](https://github.com/avoidwork/tiny-etag/compare/2.1.0...2.1.1)

> 30 December 2020

- Updating CHANGELOG.md [`d2064a5`](https://github.com/avoidwork/tiny-etag/commit/d2064a56adb9d306341b2925b0d7d673c1836778)
- Updating router dev dependency [`2b61e06`](https://github.com/avoidwork/tiny-etag/commit/2b61e06650013971c401485b90d07e604948f341)

src/constants.js
export const BASE64 = "base64";
export const SHA1 = "sha1";
export const CACHE_CONTROL = "cache-control";
export const CONTENT_LOCATION = "content-location";
export const DATE = "date";
export const ETAG = "etag";
export const EXPIRES = "expires";
export const VARY = "vary";
export const TEXT_PLAIN = "text/plain";
export const INT_DETAULT_CACHE = 1e3;
export const INT_0 = 0;
export const INT_200 = 200;
export const INT_304 = 304;
export const INT_1000 = 1000;
export const GET = "GET";
export const FINISH = "finish";
export const RANGE = "range";
export const IF_NONE_MATCH = "if-none-match";
export const EMPTY = "";
export const NO_CACHE = "no-cache";
export const NO_STORE = "no-store";
export const PRIVATE = "private";
export const PUBLIC = "public";
export const CONTENT_TYPE = "content-type";
export const MAX_AGE_0 = "max-age=0";
export const STRING = "string";
